File: 3 Year/ComputationalGeometry/convex-hull/region_tree/region_tree.py
from convex_hull.algorithm import Algorithm
from blist import sortedlist

class RegionTreeSearch(Algorithm):
    class TreeNode:
        def __init__(self, left=0, right=0):
            self.left_x = 0
            self.right_x = 0
            self.left = left
            self.right = right
            self.points = sortedlist(key=lambda point: point.y)

    class Tree:
        def __init__(self, points):

            self.points = sorted(points, key=lambda point: point.x)
            self.root = RegionTreeSearch.TreeNode()
            for point in points:
                self.root.points.add(point)

            self.root.left_x = min(self.points, key=lambda point: point.x).x
            self.root.right_x = max(self.points, key=lambda point: point.x).x + 1
            self.build_tree(self.root)

            self.left_x = None
            self.right_x = None
            self.top_y = None
            self.bottom_y = None

        def build_tree(self, node):
            points = sorted(node.points, key=lambda x: x.x)
            l = len(node.points)

            if l > 1:
                node.left = RegionTreeSearch.TreeNode()
                for point in points[: l // 2]:
                    node.left.points.add(point)
                node.right = RegionTreeSearch.TreeNode()
                for point in points[l // 2:]:
                    node.right.points.add(point)

                node.left.left_x = node.left_x
                node.right.right_x = node.right_x

                node.left.right_x = max(node.left.points, key=lambda point: point.x).x
                node.right.left_x = min(node.left.points, key=lambda point: point.x).x

                self.build_tree(node.right)
                self.build_tree(node.left)

        def search(self, node, res_list):
            if len(node.points) > 1:
                if self.left_x <= node.left.right_x:
                    self.search(node.left, res_list)
                if self.right_x >= node.right.left_x:
                    self.search(node.right, res_list)

            else:
                point = node.points[0]
                if self.left_x <= point.x < self.right_x and \
                   self.bottom_y <= point.y <= self.top_y:
                    res_list.append(point)

    def __init__(self, canvas=None, animate=False):
        super().__init__(canvas, animate)
        self.tree = None
        self.points = []

    def get_points(self, point1, point2):

        self.tree.left_x = min(point1.x, point2.x)
        self.tree.right_x = max(point1.x, point2.x)
        self.tree.top_y = max(point1.y, point2.y)
        self.tree.bottom_y = min(point1.y, point2.y)
        res = []
        self.tree.search(self.tree.root, res)
        return res

    def start(self, points):
        self.points = points
        self.tree = RegionTreeSearch.Tree(self.points)

    def draw(self, points=None, color="violet"):
        if points is None:
            return []
        res = []
        for point in points:
            res.append(point.draw(self.canvas, color=color))
        self.canvas.update()
        return res

File: 3 Year/ComputationalGeometry/GraphUtils-master/src/univgraphics/geomsearch/RegionTree.java
package univgraphics.geomsearch;

import univgraphics.common.primitives.Node;
import univgraphics.common.primitives.Point;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Ihor Handziuk on 10.04.2017.
 * All code is free to use and distribute.
 */
public class RegionTree {

    private static class TreeNode {
        int xLeft, xRight; // interval [xLeft, xRight)
        TreeNode left;
        TreeNode right;
        List<Point> points = new ArrayList<>();

    }

    private List<Node> graph;
    private int leftX, rightX;
    private int topY, bottomY;
    private TreeNode root;

    public RegionTree(List<Node> graph) {
        this.graph = graph;
        root = new TreeNode();
        graph.sort(Comparator.comparingInt(Point::getX));
        root.points.addAll(graph);
        root.xLeft = graph
                .stream()
                .min(Comparator.comparingInt(Point::getX))
                .get()
                .getX();
        root.xRight = graph
                .stream()
                .max(Comparator.comparingInt(Point::getX))
                .get()
                .getX() + 1;  // + 1 for maintaining the rule [xLeft, xRight)
        buildTree(root);
    }

    public List<Point> getPoints(Point startCorner, Point endCorner) {
        leftX = Math.min(startCorner.getX(), endCorner.getX());
        rightX = Math.max(startCorner.getX(), endCorner.getX());
        topY = Math.max(startCorner.getY(), endCorner.getY());
        bottomY = Math.min(startCorner.getY(), endCorner.getY());

        List<Point> res = new ArrayList<>();
        search(root, res);
        return res;
    }

    private void search(TreeNode parent, List<Point> list) {
        if (parent.points.size() > 1) {
            if (leftX <= parent.left.xRight) {
                search(parent.left, list);
            }
            if (rightX >= parent.right.xLeft) {
                search(parent.right, list);
            }
        } else {
            Point point = parent.points.get(0);
            if (point.getX() > leftX &&
                point.getX() < rightX &&
                point.getY() > bottomY &&
                point.getY() < topY) {
                list.add(point);
            }
        }
    }

    private void buildTree(TreeNode parent) {
       if (parent.points.size() > 1) {
           parent.left = new TreeNode();
           parent.right = new TreeNode();
           parent.left.points.addAll(parent.points
                   .stream()
                   .limit(parent.points.size() / 2)
                   .collect(Collectors.toList()));
           parent.right.points.addAll(parent.points
                   .stream()
                   .skip(parent.points.size() / 2)
                   .collect(Collectors.toList()));
           parent.left.xLeft = parent.xLeft;
           parent.left.xRight = parent.left.points.get(parent.left.points.size() - 1).getX();
           parent.right.xLeft = parent.right.points.get(0).getX();
           parent.right.xRight = parent.xRight;
           buildTree(parent.left);
           buildTree(parent.right);
       }
    }
}
